### PR TITLE
Add preview pane to auto currency switch feature

### LINKED ISSUE #2523: Preview for automatic currency update alert
Follow up from #2258 

We should add a preview link to the `enable_auto_currency` setting description, which allows the merchant to preview the notice on their site.

### PR DESCRIPTION
**Pull request by tpaksu**
Fixes #2523

#### Changes proposed in this Pull Request

This PR adds a preview link to Store settings section of Multi-currency settings page, which enables the user to preview how the automatic currency switch feature will affect the customer. 

**Note:** To get this work, merchant must have the local currency enabled already, because it won't show a auto currency change notice under the page when the currency isn't available. 

The preview modal is extracted from MC onboarding to its own component and rendered on the store settings section in a placeholder that's provided in the `ìncludes/multi-currency/Settings.php` settings page constructor array.

##### The link:

image

##### The modal:

image

#### Testing instructions

* Open *WooCommerce > Settings > Multi-Currency*
* Add some currencies if the list is empty to get the store settings section displayed.
* Make sure your list doesn't contain your local currency at first
* Check if the preview link is visible in the help text of Auto currency switch checkbox
* Click the link and see the notice isn't there.
* Add your currency to the list now
* Click the link and see the notice is visible, and showing your local currency is effective, check the prices are displayed in your local currency.

-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Added testing instructions to the [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) (or does not apply)

### CODE REVIEW

**Comment by dmallory42** [comment]
Thanks for working on this @tpaksu! It's looking really good.

It worked as expected when I was testing with my store currency set to USD. However, I tested changing my store currency to the same as my local currency (GBP) and the banner didn't appear in the preview. Is this expected? My worry is that the majority of merchants will have the same default store currency as their local currency (e.g. a US-based merchant will have dollars as their store currency), and therefore won't see how the notice will display / how the auto switching logic works when they open the preview pane. 

What do you think? I'm wondering whether we could just pass some test/simulated values to the preview modal to use for the preview notice rather than trying to pick up the customer's actual location/currency, since the preview is mainly just meant to show them how this may look/function on their site.

**Comment by tpaksu** [comment]
Hi @dmallory42, thanks for the review!

> I tested changing my store currency to the same as my local currency (GBP) and the banner didn't appear in the preview. Is this expected? 

I think the banner only displays when an automatic currency switch based on customer's geolocation data has happened, and thus informing the user with the change. I don't have an idea if we should simulate it in the modal, this is something that @LevinMedia should decide. Also, thanks for raising this concern, that was one of my concerns too! 

Also the auto currency switch, which works in the iframe, doesn't work in the real site. I don't know why :)

**Comment by LevinMedia** [comment]
>I don't have an idea if we should simulate it in the modal, this is something that @LevinMedia should decide.

@tpaksu I think we should go ahead and simulate this for the merchant. Can we automatically switch them to the first currency in this list, and show them the notice for that in the preview?

**Comment by tpaksu** [comment]
@LevinMedia sure, I can do that.

**Comment by LevinMedia** [comment]
@tpaksu one more thing I noticed in testing, when I'm not using store front, the notice about the currency switcher in your widgets is appearing outside the card, rather than under the preview notice - not sure if that's directly related to this issue. If it's not I can open another issue for it, or if it's not too much trouble perhaps you could wrap it up in this PR too. 

Screen Shot 2021-09-17 at 2 38 26 PM

**Comment by tpaksu** [comment]
@dmallory42 @LevinMedia I've updated the code to always show the notice for the first enabled currency in the preview iframe. Also put the string in it's appropiate place when the theme is not Storefront. You can review the code and the design again. Thanks in advance!

**Comment by oaratovskyi** [comment]
Same for me in case my local currency is enabled.
image

**Comment by LevinMedia** [comment]
@tpaksu I think we should pair a proper country with the currency we're switching to... in the case of multiple countries using a single currency, is there anyway we can leverage that cldr data to grab the first applicable country for a currency?

**Comment by tpaksu** [comment]
@LevinMedia I thought that too but selecting a country for international currencies seems not possible easily. It might cause confusions like showing Austria in France for EUR for example.

**Comment by tpaksu** [comment]
@LevinMedia I have an idea about that, how about adding a country selector combo as a toolbar in the preview modal, just above the iframe, and let the user select the country?

File: client/multi-currency/index.js
Comment (oaratovskyi) [inline client/multi-currency/index.js:183]: I am curious why don't we need this switcher in preview too?

File: client/multi-currency/index.js
Comment (oaratovskyi) [inline client/multi-currency/index.js:184]: Just leaving a comment here till the first currency approach is finalized or not.

File: includes/multi-currency/MultiCurrency.php
Comment (oaratovskyi) [inline includes/multi-currency/MultiCurrency.php:1227]: I suggest adding unit tests for this method, since the logic is not trivial

File: includes/multi-currency/MultiCurrency.php
Comment (oaratovskyi) [inline includes/multi-currency/MultiCurrency.php:1200]: I am not sure that I understand how that works, the `$selected_currency` variable is not used here. 
Would you explain it a bit to me, please?

File: includes/multi-currency/MultiCurrency.php
Comment (oaratovskyi) [inline includes/multi-currency/MultiCurrency.php:613]: not sure but probably existing unit test should be modified to cover this case

File: includes/multi-currency/MultiCurrency.php
Comment (oaratovskyi) [inline includes/multi-currency/MultiCurrency.php:1206]: Just a quick question: would it conflict with line 677?
https://github.com/Automattic/woocommerce-payments/pull/2923/files#diff-c722c76bb4254c6fe1a32099d4ea5fa641addb44c5cd685a164e8baaff879e1aR677

File: includes/multi-currency/MultiCurrency.php
Comment (oaratovskyi) [inline includes/multi-currency/MultiCurrency.php:1164]: [minor] I suggest renaming it to `validate_and_transform_currency_code` since not only validation is done

**Review by oaratovskyi — COMMENTED** [review]
@tpaksu it looks really good! 💪
I've left a couple of questions and suggestions.
The main point from me - to cover it a bit more with unit tests.
Let's wait for David's opinion on the first currency moment and I'll review it and test one more time.

File: client/multi-currency/index.js
Comment (tpaksu) [reply client/multi-currency/index.js:183]: image

The *Preview* link is bound to the auto currency switch feature, and I thought that it will only provide the preview for that feature only. In the onboarding, the link was near the `Continue` button and it provided the preview for both features.

File: includes/multi-currency/MultiCurrency.php
Comment (tpaksu) [reply includes/multi-currency/MultiCurrency.php:1200]: I'm not sure which part you didn't understand, but sure, I'll explain both the possibilities. 

### 1. You didn't understand how the filter works

The filter provides the current value (`$selected_currency`) as the parameter for us to tinker on it and return the modified version back, thats how `add_filter` works, but we don't use that and return our `$simulation_currency` instead. 

I'll explain with an example.

```php 

$p = 2;

$p = apply_filters(`p_value_modifiers`, $p);

echo $p;
````

If there's no filters hooked to the `p_value_modifiers` hook, it'll echo 2. If we add this:

```php
add_filter(`p_value_modifiers`, function($old_p){ return $p * 2; });
```

it'll echo 4, because the $p will be modified with that filter.

But here, we're not modifying the data, we're replacing it, so think it like this:

```php
add_filter(`p_value_modifiers`, function($old_p){ return 5; });
```

it'll echo 5.

### 2. You didn't understand why we keep the `$selected_currency` as a parameter and gave the variable as an outside variable;

The filters are probably applied with `call_user_func`, and that'll expect the callback to have the matching parameter count with the filter itself. Otherwise it'll throw an error.

File: includes/multi-currency/MultiCurrency.php
Comment (tpaksu) [reply includes/multi-currency/MultiCurrency.php:1206]: Don't know, I didn't think about it since you pointed it out here :) You're right, the filter might've been registered twice, then there should be two notice's overlapped. I'll check in the markup if it works that way. Thanks for pointing that out!

File: includes/multi-currency/MultiCurrency.php
Comment (tpaksu) [reply includes/multi-currency/MultiCurrency.php:1164]: I think *transform* means something that provides a bigger change than converting the case to match the keys in the arrays here, so I'll probably pass this. How does it sound?

File: includes/multi-currency/MultiCurrency.php
Comment (tpaksu) [reply includes/multi-currency/MultiCurrency.php:1227]: How about converting it to `private`?

**Comment by LevinMedia** [comment]
@tpaksu I feel like adding a select combo to this is overkill - what if we just picked a country / currency pair and always used that, regardless of whether or not the currency is active in the store. Perhaps a more obscure country currency combo?

**Comment by tpaksu** [comment]
Sure, that'll work. We can do this with or without the combo. What do you think? Two country currency pairs to switch between if one is the store currency, or 4-5 pairs which the user can select and simulate? I'd prefer the first one.

File: client/multi-currency/index.js
Comment (oaratovskyi) [reply client/multi-currency/index.js:183]: Gotcha, thanks.

File: includes/multi-currency/MultiCurrency.php
Comment (oaratovskyi) [reply includes/multi-currency/MultiCurrency.php:1227]: That works, would be great to have some high-level test though that covers that private method 👍

File: includes/multi-currency/MultiCurrency.php
Comment (oaratovskyi) [reply includes/multi-currency/MultiCurrency.php:1164]: Reasonable.
Yeah, it's better to leave it than renaming to something like `validate_and_uppercase`

File: includes/multi-currency/MultiCurrency.php
Comment (oaratovskyi) [reply includes/multi-currency/MultiCurrency.php:1200]: Thanks for a detailed explanation, I really appreciate it! 

I meant 

> 2. You didn't understand why we keep the $selected_currency as a parameter and gave the variable as an outside variable;

Yes, they are applied with `call_user_func `, I've re-read the code once again, now it's clear 👍

File: includes/multi-currency/MultiCurrency.php
Comment (tpaksu) [reply includes/multi-currency/MultiCurrency.php]: Fixed.

File: includes/multi-currency/MultiCurrency.php
Comment (tpaksu) [reply includes/multi-currency/MultiCurrency.php:613]: Fixed.

File: includes/multi-currency/MultiCurrency.php
Comment (tpaksu) [reply includes/multi-currency/MultiCurrency.php:1232]: Fixed.

**Comment by LevinMedia** [comment]
@tpaksu let's go with your first suggestion:

> Two country currency pairs to switch between if one is the store currency,

**Comment by tpaksu** [comment]
> let's go with your first suggestion

Added *USD => United States (US)* as the first option and *GBP => United Kingdom (UK)*  as a fallback.

File: client/multi-currency/index.js
Comment (tpaksu) [reply client/multi-currency/index.js]: Updated the code on 4ffab70, removed all instances of `clientCurrency` and moved the simulation currency selection to PHP side. https://github.com/Automattic/woocommerce-payments/blob/4ffab7001b0e47f3107ecd4d5d7b60773bbf1f0b/includes/multi-currency/MultiCurrency.php#L1199-L1207

**Review by oaratovskyi — APPROVED** [review]
@tpaksu Thank you for addressing the PR feedback, for your work on tests and on currencies that are simulated. Now it really looks good to me!
One small thing that I noticed: if we have just one currency which is store currency, the alert banner is still showing up. Basically it makes sense because we say `Customers will be notified via store alert banner `. Technically the customer might want to add more currencies after sees the banner. @LevinMedia does this sound good to you too?

I'm pre-approving it, feel free to merge after the final design review 🚢

**Comment by LevinMedia** [comment]
Thanks @tpaksu this looks good to me, my only comment would be to preview the alert over the shop page rather than the home page (the setting for the shop page is in the products section of the WooCommerce settings. If a shop page isn't set, then we can fall back to the home page.

**Comment by oaratovskyi** [comment]
Reviewed and tested last changes.
Great work :shipit:

**Pull request merged**
